internal/graph/check.go

```diff
@@ -768,7 +768,13 @@ func (c *LocalChecker) checkDirect(parentctx context.Context, req *ResolveCheckR
 		resp, err := union(ctx, c.concurrencyLimit, checkFuncs...)
 		if err != nil {
 			telemetry.TraceError(span, err)
+			return nil, err
 		}
+		if len(directlyRelatedUsersetTypes) > 0 {
+			// if we had N userset checks, that was 1 read, not N
+			resp.GetResolutionMetadata().DatastoreQueryCount++
```

Comment:
Feels a bit weird to add it here rather than closer to the source we call the datastore query

Reply:
I agree, but i don't know how to make it better. This is counting + 1 in the case where we didn't find a path that led to `allowed=true`.

Reply:
I think it would help composition and readability here to keep track of parent and child aggregates independently, so we don't stomp on each other. A single parent spawns multiple child userset evals, but they are overlapping on a shared resolution metadata field.

What we probably should do is the following:
```go
type ResolveCheckRequest struct {
  ParentRequest *ResolveCheckRequest
  ...
  ResolutionMetadata ResolveCheckRequestMetadata
}

type ResolveCheckRequestMetadata struct {
   DatastoreQueryCount int
}

func ... checkDirect(..., req *ResolveCheckRequest) {
    fn2 := ... {
      iter, err := ds.ReadUsersetTuples(...)
      if err != nil {
        // handle error
      }

      req.GetResolutionMetadata().DatastoreQueryCount += 1

      for {
        iter = iter.Next()
        ...

        if usersetRelation != "" {
          ...

          handlers = append(handlers, c.dispatch(
            ctx,
            &ResolveCheckRequest{
              ParentRequest: req,
              ...
              ResolutionMetadata: &ResolutionMetadata{
                Depth:               req.GetResolutionMetadata().Depth - 1,
                DatastoreQueryCount: 0, // start at 0 for the subproblem, we'll aggregate as we query
              },
              ...
            }))
        }
      }

      resp, err := union(ctx, c.concurrencyLimit, handlers...)
      if err != nil {
        ...
      }

      // req is the parent, remember
      // add the aggregate sum of the children datastore query counts, which are aggreated
      // in the response.
      req.GetResolutionMetadata().DatastoreQueryCount += resp.GetResolutionMetadata().DatastoreQueryCount
    }
}
```

Reply:
> The datastore query count increment should be part of fn2 (only when fn2 is invoked - as in the case of line 765), the datastore query count is incremented.

we can't do that because the outcome is incorrect.

- `union(allowed=true, allowed=false)` should return the result of the reads for `allowed=true` branch only, which we _can_ count inside `fn2`
- however, `union(allowed=false, allowed=false)` should return the sum of _all_ reads that we had to do. if we add +1 to each individual branch, we will be over-counting

Reply:
Tracking original comment in https://github.com/openfga/openfga/issues/1477